docs/content/docs/(root)/guides/model-context-protocol.mdx

```diff
@@ -43,9 +51,147 @@ export default function PageLevelComponent() {
 }
 ```
 
-### Backend Setup
+## Rendering MCP Tool Calls
 
-Configure your backend to handle MCP requests (this example uses Next.js App Router):
+You can visualize MCP tool calls in your UI by using the `useCopilotAction` hook with a catch-all action handler:
+
+```tsx
+"use client";
+
+import {
+  useCopilotAction,
+  CatchAllActionRenderProps,
+} from "@copilotkit/react-core";
+import MCPToolCall from "./MCPToolCall";
+
+export function ToolRenderer() {
+  useCopilotAction({
+    name: "*", // Catch all tool calls
+    render: ({ name, status, args, result }: CatchAllActionRenderProps<[]>) => (
+      <MCPToolCall status={status} name={name} args={args} result={result} />
+    ),
+  });
+  return null;
+}
+```
+
+### Sample MCPToolCall Component
+
+<details>
+<summary>Click to view the MCPToolCall component implementation</summary>
+
+```tsx
+"use client";
+
+import * as React from "react";
+
+interface ToolCallProps {
+  status: "complete" | "inProgress" | "executing";
+  name?: string;
+  args?: any;
+  result?: any;
+}
+
+export default function MCPToolCall({
+  status,
+  name = "",
+  args,
+  result,
+}: ToolCallProps) {
+  const [isOpen, setIsOpen] = React.useState(false);
+
+  // Format string with proper line breaks and quotes
+  const formatContent = (str: string): string => {
+    const text = typeof str !== "string" ? JSON.stringify(str, null, 2) : str;
+    return text
+      .replace(/\\n/g, "\n")
+      .replace(/\\t/g, "\t")
+      .replace(/\\"/g, '"')
+      .replace(/\\\\/g, "\\");
+  };
+
+  // Format arguments and results
+  const formattedArgs = React.useMemo(() => {
+    if (!args) return "";
+    return formatContent(
+      typeof args === "object" ? JSON.stringify(args, null, 2) : args
+    );
+  }, [args]);
+
+  const formattedResult = React.useMemo(() => {
+    if (!result) return "";
+
+    if (typeof result === "object" && result?.error?.message) {
+      const copy = { ...result };
+      copy.error = { ...copy.error };
+      copy.error.message = formatContent(copy.error.message);
+      return formatContent(JSON.stringify(copy, null, 2));
+    }
+
+    return formatContent(
+      typeof result === "object" ? JSON.stringify(result, null, 2) : result
+    );
+  }, [result]);
```

Comment:
This is a little complex, we should try simplifying this down. Basically, how can we compress this to maximize user understanding when they see this.